solr/solrj/src/java/org/apache/solr/common/util/CommonTestInjection.java

```diff
@@ -75,4 +78,49 @@ public static boolean injectDelay() {
     }
     return true;
   }
+
+  /**
+   * This should ONLY be set from unit test cases.
+   *
+   * Block at certain code execution point (defined by the key at which injectWaitForCondition was inserted)
+   * until the provided condition is fulfilled
+   * 
+   * @see CommonTestInjection#injectWaitForCondition(String)
+   * @param key
+   * @param condition
+   */
+  public static void setWaitForCondition(String key, Condition condition) {
+    synchronized (waitForConditions) {
+      if (condition != null) {
+        waitForConditions.put(key, condition);
+      } else {
+        waitForConditions.remove(key);
+      }
+    }
+  }
+
+  /**
+   * This injects a precise wait until a condition is fulfilled, such condition is looked up by the corresponding key.
+   *
+   * The key could simply be the fully qualified class name or could be more granular like class name + other id
+   * (such as method name) and should only be set by corresponding unit test cases with setWaitForCondition
+   *
+   * Therefore, in non-testing environment, this should always return w/o blocking since the lookup map should be empty.
+   *
+   * @see CommonTestInjection#setWaitForCondition(String, Condition) 
+   * @param key
+   */
+  public static void injectWaitForCondition(String key) {
```

Comment:
I think we should always return `true` here, as with `injectDelay()`. This makes it possible to invoke these methods via standard java assertions, thus ensuring that they don't add any overhead unless assertions are enabled (which they are in tests).

Reply:
ah that's a great idea! I was a bit disturbed by the fact that unit test related code appears in regular code. But guarding behind an assert does make it much less intrusive (though we still see the code there...)

Reply:
Yeah -- this approach also makes it so that no matter how complex the code you inject (of course ideally it's pretty straightforward), the additional complexity is in the test code, and main sources only get stubs (which are even arguably helpful wrt transparency, and succinct enough to not pull focus from the main logic of the program).

(btw I can't take credit for the "hide behind an assert" approach -- just following precedent; but I agree it's a nice approach!)